lib/srv/desktop/session.go

```diff
@@ -0,0 +1,186 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package desktop
+
+import (
+	"context"
+	"sync"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/auth"
+	"github.com/gravitational/teleport/lib/services"
+	sshsession "github.com/gravitational/teleport/lib/session"
+	"github.com/gravitational/teleport/lib/srv/desktop/rdp/rdpclient"
+	"github.com/gravitational/teleport/lib/srv/desktop/tdp"
+	"github.com/gravitational/trace"
+
+	"github.com/sirupsen/logrus"
+)
+
+// newSession creates a new session object, waits for initial TDP messages,
+// checks that the user has permission to create a new session with the requested desktop,
+// and then if everything checks out, creates the RDP connection.
+func newSession(ctx context.Context, authCtx *auth.Context, desktop types.WindowsDesktop, sessionID sshsession.ID, tdpConn *tdp.Conn, rdpc *rdpclient.Client, log logrus.FieldLogger) (*session, error) {
+	s := &session{
+		sessionID: sessionID,
+		tdpConn:   tdpConn,
+		rdpc:      rdpc,
+		log:       log,
+	}
+
+	u, err := s.waitForUsername()
+	if err != nil {
+		return nil, err
+	}
```

Comment:
@zmb3 i am curious why sending username and screen size are asynchronous operations?

Reply:
@alex-kovoy I'm not exactly sure what you mean by asynchronous operations in this context. 

We receive those from the websocket string [here](https://github.com/gravitational/teleport/blob/e3fa4d7f00445c1e4149ff82090820b38117ad75/lib/web/desktop.go#L73-L85) in the `proxy` and then [dial out](https://github.com/gravitational/teleport/blob/e3fa4d7f00445c1e4149ff82090820b38117ad75/lib/web/desktop.go#L109) to the `windows_desktop_service`, which is separate from the `proxy`. We need somehow pass them to that service, so we [send them as TDP messages](https://github.com/gravitational/teleport/blob/e3fa4d7f00445c1e4149ff82090820b38117ad75/lib/web/desktop.go#L125-L134). That function is the other side of that connection, waiting to read the username and size.

Reply:
> That function is the other side of that connection, waiting to read the username and size.

@ibeckermayer  Yeah, i am familiar with the flow but my question is why do we need to send username and screen size and connection requests _separately_ to establish a connection? Just trying to understand the reason behind this.

Reply:
Here are 2 questions.

1. Can we pass this information together with the servername as part of tlsConfig? That would be ideal from API perspective. 
2. If not, instead of using 2 messages that we sent one by one, can we combine it in 1 with a dedicated message type?

```golang
tdpConn := tdp.NewConn(serviceConTLS)
	err = tdpConn.OutputMessage(tdp.StartSession{
		User: user
		DisplayHeight: 330,
		DisplayWidth: 200	
	})
```

Reply:
1. I had a version that did that, but you end up with a gross implicit api and string parsing mess [on the other end](https://github.com/gravitational/teleport/pull/9088/files#diff-d0d78875d7396d7dad9cc4ab10f6da26554c8e8ff30f2b661c1377d9b959bd85). Using the SNI is a hack in the first place and I agree with @zmb3 's assessment of that PR, which is that the hack shouldn't be extended.
2. We could modify the TDP spec to have that message instead. Then would could remove the [`client username`](https://github.com/gravitational/teleport/blob/master/rfd/0037-desktop-access-protocol.md?plain=1#L175-L182) message and rename [`client screen size`](https://github.com/gravitational/teleport/blob/master/rfd/0037-desktop-access-protocol.md?plain=1#L98-L108) to `resize`.
 
I think (2) is a better design than what we have now, because its clear that `StartSession` is meant to be sent before any other tdp message. Currently, its just implicit in the implementation that a TDP session is started by sending the username and then screen spec.

Reply:
1. I would use the standard URI to represent the conn string, this would have simplified parsing (if that was the main issue).
2. If `1` is not an option, I think we need to simplify it at least with `StartSession` message and update the specs (which is still beta)

Reply:
1. Do you mean put it [here](https://github.com/gravitational/teleport/blob/e3fa4d7f00445c1e4149ff82090820b38117ad75/lib/web/desktop.go#L111)? The other side is [this plain `Listener`](https://github.com/gravitational/teleport/blob/e3fa4d7f00445c1e4149ff82090820b38117ad75/lib/srv/desktop/windows_server.go#L543), is it possible to get URI parameters from it?
2. If so, `1` is an option. But would it truly be a better option? It seems to me that TDP would be a more complete protocol with a `StartSession` message.

Reply:
I would discuss it with @zmb3 to see what the the best option, yeah `2` seems fine too.

Reply:
>  They're not asynchronous operations, they just need to be sent before the backend initiates the RDP connection to the desktop (as these parameters are used to establish the connection)

@zmb3 what I meant to say to replace the following:

```golang
u, err := s.waitForUsername()
if err != nil {
        return nil, err
}

sc, err := s.waitForClientSize()
if err != nil {
	return nil, err
}
```
	
With something like 

```golang
, err := s.waitForSessionStart()
if err != nil {
	return nil, err
}
```

Reply:
Yes, we could combine the username and screensize into a single TDP message rather than two separate messages. For RDP this makes a lot of sense and probably would have been a better design from the beginning.

Unfortunately, we've already released software that uses the existing TDP messages, so I can't remove them. The best I could do is add a new message and slowly migrate to it while retaining compatibility for the old message.

Given all the work we have left to do on desktop access, I think creating a new message and maintaining two sets of nearly identical messages is just extra work that doesn't provide any major advantages and creates more code for us to maintain.

Once we get through the remaining features required for Teleport 9 and get through the fuzzing and vulnerability testing of the TDP protocol I am happy to review the protocol in its entirety and see if we want to plan any revisions for Teleport 10.

Sound reasonable?